### DISCUSSION TITLE
BPF rules don't block any traffic on Suricata, Zeek and Steno

### DISCUSSION THREAD

**Discussion by Axtrus** [Security-Onion-Solutions/securityonion#11625] [2.4]
### Version

2.4.20

### Installation Method

Security Onion ISO image

### Description

other (please provide detail below)

### Installation Type

Standalone

### Location

on-prem with Internet access

### Hardware Specs

Exceeds minimum requirements

### CPU

64

### RAM

252

### Storage for /

2 TB

### Storage for /nsm

2 TB

### Network Traffic Collection

span port

### Network Traffic Speeds

1Gbps to 10Gbps

### Status

Yes, all services on all nodes are running OK

### Salt Status

No, there are no failures

### Logs

No, there are no additional clues

### Detail

Hello everyone,

I'd like some help on my security onion server, in fact I'm trying to filter certain traffic in my network that I found to be noisy.
I've followed the instructions on the Security Onion documentation and created the following rules which are applied to the config section of BPF suricata, zeek and PCAP:

not (host 172.16.12.10 and host 172.30.224.94) &&
not (host 170.32.24.243 or host 170.41.0.67 or host 170.32.3.213 or host 170.34.3.42) &&
not (host 173.41.0.100 or host 173.34.0.110 or host 173.32.2.39 or host 173.32.1.110 or host 173.32.224.110) &&
not tcp dst port 7081 &&
not (src port 0 or dst port 0)

None of the above rules worked as intended, I'm still seeing Suricata alerts from these hosts. 

Does anyone know how to make it work to block traffic from all hosts in the BPF rules above?  
And could be the problem why it's not working ?

Regards.

### Guidelines

- [X] I have read the above statement and can confirm my post is relevant to Security Onion 2.4.

**Comment by dougburks** [comment]
Were you following the 2.4 documentation?
https://docs.securityonion.net/en/2.4/bpf.html

**Reply by Axtrus** [reply]
Yes Doug

**Reply by dougburks** [reply]
If you simplify the BPF and just specify one host like the following, does that work?
```
not host 172.16.12.10
```
Keep in mind that, depending on your configuration, it may take 15-30 minutes for the new configuration to take effect.

**Reply by Axtrus** [reply]
I already did that and waited for an hour for the configurations to take effect, but it didn’t work.

**Reply by dougburks** [reply]
Can you share a screenshot of where exactly you're placing that config?

Can you share a second screenshot of the alerts or logs that appeared after placing the config that you're wanting to get rid of?

**Reply by Axtrus** [reply]
Hi Doug

Here are the screenshots requested, I've just hidden the true IPs.
The configs are in SOC->Configuration -> BPF -> Suricata/Pcap/Zeek -> seconion_standalone(Grid Value)

![BPF- Second Alert example to filter based on the host](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/b76529a0-4baf-41d5-bd9a-fa55ca0410c2)
![BPF-Alert example to filter](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/6c0d350b-e343-4873-90f8-9a863a333eaa)
![BPF-Config](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/93ca95d0-4cfe-44df-b8f9-a46c8cf71bea)

**Reply by dougburks** [reply]
Since we're troubleshooting, we need to remove as many variables as possible. As I mentioned previously, please simplify your suricata BPF to just the following and nothing else:
```
not host 172.16.12.10
```
Then run the following:
```
sudo salt-call state.highstate
```
Assuming there are no errors, check to see if you continue to receive new alerts from host 172.16.12.10. If so, please provide screenshots of all of the above.

**Reply by Axtrus** [reply]
Hi Doug
I followed your suggestion and did it, but it still doesn't work at all. 
As you can see in the image below, there were no errors.
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/e8a2b060-2314-4dba-a896-a19f32e00244)

**Reply by dougburks** [reply]
As I mentioned in my last reply, please provide screenshots of all of the above (meaning show the BPF config and the new alerts that occurred after the highstate was complete and the BPF was fully active). Also show the content of `/opt/so/conf/suricata/bpf`.

**Reply by Axtrus** [reply]
Here are the requested screenshots 

![image](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/0172ba50-bafe-467f-96f3-b216145806ce)
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/f5f2cb4b-0217-402e-9d90-3ccb38e0b541)
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/cac986e9-f1d3-4301-a879-d5ea9be78bfa)

**Reply by dougburks** [reply]
This is strange as I've tested Suricata BPFs on a system here and it works fine for me.

Can you confirm that this is a standalone installation and there are no other sensors sending Suricata alerts?

Have you made any other changes to the configuration?

What is the output of the following?
```
ps aux |grep suricata
```

Have you tried restarting Suricata?

Have you tried rebooting the box?

**Reply by Axtrus** [reply]
I confirm that this is a standalone installation and i have made no changes to the configuration 
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/ae855a2f-d551-45b5-8ab7-c2bc081e1534)

- Yes I've already restarted Suricata but it didn't work.

Yesterday I've rebooted the box after updating the BPF rules on Suricata and Steno. To my big surprise I saw that almost all rules are working now except this : **not (src port 0 or dst port 0)** as you can see in the screenshot below
![image](https://github.com/Security-Onion-Solutions/securityonion/assets/74593771/d4507909-7d43-4346-b142-e4af7874fd9c)

**Reply by dougburks** [reply]
Looking at the screenshot, the alerts are for ICMP traffic. Unlike TCP and UDP, ICMP doesn't really have any concept of source port or destination port. (Suricata logs src_port as `0` and dest_port as `0` simply to indicate that there was no actual port value.) Therefore, applying a port-based BPF won't have any effect on ICMP traffic. Here are some alternatives:
- you could apply a BPF to exclude ICMP traffic altogether or based on source IP address or destination IP address (or both)
- you could disable or suppress the alert itself as shown at https://docs.securityonion.net/en/2.4/managing-alerts.html#disable-the-alert

**Reply by Axtrus** [reply]
Okay well noted, thanks for your feedback and support Doug.

**Comment by InfosecGoon** [comment]
It looks like you're using a mix of ANDs and ORs, which is a little confusing and may not be applying in the way you intend. Try something like this to ignore traffic from those hosts:

not host 172.16.12.10 && 
not host 172.30.224.94 &&
not host 170.32.24.243 &&
not host 170.41.0.67 &&
not host 170.32.3.213 &&
not host 170.34.3.42 &&
not host 173.41.0.100 &&
not host 173.34.0.110 &&
not host 173.32.2.39 && 
not host 173.32.1.110 &&
not host 173.32.224.110

**Comment by Axtrus** [comment]
Thanks for your response, I changed the configs and applied the rules as you've suggested but i got the same results

**Comment by Acewiza** [comment]
I've had issues with the BPF stuff.  Not offering any insight or suggestions, but I've struggled with it in the past, not attempting it again until something changes.

LATE UPDATE:  I take it back - working for me now.

**Comment by petiepooo** [comment]
Here's how we format and validate our BPFs; YMMV.

We format the yaml like this with different global and site-specific exclusions noted as comments:
```
zeek:
  bpf:
    - "!(udp port 123)&&" # GLOBAL
    - "!(ip net 15.0.0.0/8)&&" # HP GLOBAL
    - "!(ip net 17.0.0.0/8)&&" # APPLE GLOBAL
    - "!(ip net 31.13.74.0/24)&&" # FACEBOOK GLOBAL
    - "!(ip net 157.240.0.0/16)" # FACEBOOK GLOBAL
```
    
We also have installed the 'yq' package on the system, which is a yaml parsing and analysis tool similar to what jq is for json.  With that tool, we are able to run this command to verify that the yaml syntax is correct and to generate the resulting concatenated BPF expression:

> \# yq '.zeek.bpf | ... comments="" | join(.)' /opt/so/saltstack/local/pillar/global.sls 
!(udp port 123)&&!(ip net 15.0.0.0/8)&&!(ip net 17.0.0.0/8)&&!(ip net 31.13.74.0/24)&&!(ip net 157.240.0.0/16)

You can run the resulting expression into tcpdump in debug mode to verify it is valid BPF:

> \# tcpdump -d '!(udp port 123)&&!(ip net 15.0.0.0/8)&&!(ip net 17.0.0.0/8)&&!(ip net 31.13.74.0/24)&&!(ip net 157.240.0.0/16)'
(000) ldh      [12]
(001) jeq      #0x86dd          jt 2	jf 8
(002) ldb      [20]
(003) jeq      #0x11            jt 4	jf 43
(004) ldh      [54]
(005) jeq      #0x7b            jt 42	jf 6
(006) ldh      [56]
(007) jeq      #0x7b            jt 42	jf 43
.........

You can use the so-bpf-compile script to check validity of the bpf syntax, which just calls tcpdump behind the scene and outputs the compiled BPF in bytecode:

> \# so-bpf-compile eno2 '!(udp port 123)&&!(ip net 15.0.0.0/8)&&!(ip net 17.0.0.0/8)&&!(ip net 31.13.74.0/24)&&!(ip net 157.240.0.0/16)'

Note the use of 'ip' preceeding each net or host block in the example.  Without that, the block is checked twice: as bare IP and with a VLAN header.  If you know you will not receive any VLAN headers, adding IP as part of the expression simplifies the resulting filter by avoiding that second check.  To see for yourself, check the resulting debug output from tcpdump with or without it.

Also note extensive use of '&&' for AND and '!' for NOT.  There is a limit to the length of the BPF string before zeek barfs on it, and those tokens are shorter than their equivalent strings, plus they allow you to eliminate surrounding spaces when placed next to other strings.  Eg. you can use '!(ip&&port 53)' instead of '!(ip AND port 53)' which is a 12% reduction in string length for that example.  

There is also a maximum length for the compiled BPF string that stenographer will accept.  I have not seen these string length limits documented and have not fully characterized them myself, so just beware of really long BPFs.

**Reply by Axtrus** [reply]
Hi thanks so much for your participation and feedback but i tried to follow your suggestion and it didn't work.... I'm still seeing the traffic i don't want to see

**Comment by Acewiza** [comment]
You might actually want to "see" those alerts if using the log4 vulnerabilities - just not from certain hosts.  Try putting just one argument in the 1st dialog and select the desired onion below, Like Doug suggested.

**Comment by innovate-support** [comment]
I experienced the same issue with BPF not seeming to work at all, and I was only trying to ignore 1 IP.
Only a reboot of the Sensor resolved the issue.  After the reboot all the alerts that I wanted to ignore stopped appearing as expected.
I'm using v2.4.30 in a Distributed deployment.
